File: dart/planet-time/lib/src/time_calc.dart
// time_calc.dart — getPlanetTime and getMtc
// Ported verbatim from planet-time.js (Story 18.12)

import 'constants.dart';
import 'models.dart';

// ── Planet time ───────────────────────────────────────────────────────────────

/// Get the local time on [planet] at [utcMs].
///
/// [tzOffsetH] is the optional zone offset in planet local hours from the
/// planet's prime meridian (e.g. +4 for AMT+4 on Mars).
PlanetTime getPlanetTime(Planet planet, int utcMs, {double tzOffsetH = 0.0}) {
  // Moon uses Earth's solar day (tidally locked; work schedules run on Earth time)
  final effective = (planet == Planet.moon) ? Planet.earth : planet;
  final pd = planetDataMap[effective]!;

  final solarDay = pd.solarDayMs.toDouble();
  final elapsedMs = (utcMs - pd.epochMs).toDouble() + tzOffsetH / 24.0 * solarDay;
  final totalDays = elapsedMs / solarDay;
  final dayNumber = totalDays.floor();
  final dayFrac = totalDays - dayNumber;

  final localHour = dayFrac * 24.0;
  final h = localHour.floor();
  final minF = (localHour - h) * 60.0;
  final m = minF.floor();
  final s = ((minF - m) * 60.0).floor();

  // Work period (positive modulo so pre-epoch dates return valid range)
  final int piw;
  final bool isWorkPeriod;
  final bool isWorkHour;
  if (pd.earthClockSched) {
    // Mercury/Venus: solar day >> circadian rhythm; use UTC Earth-clock scheduling
    // UTC day-of-week: ((floor(utcMs / 86400000) % 7 + 10) % 7 → Mon=0..Sun=6
    // (+7 before +3 ensures positive result for pre-1970 timestamps)
    final utcDay = (utcMs / 86400000.0).floor();
    piw = ((utcDay % 7 + 10) % 7).toInt();
    isWorkPeriod = piw < pd.workPeriodsPerWeek;
    // UTC hour within the day — positive modulo handles pre-1970 timestamps
    final msInDay = ((utcMs % 86400000) + 86400000) % 86400000;
    final utcHour = msInDay / 3600000.0;
    isWorkHour = isWorkPeriod && utcHour >= pd.workStart && utcHour < pd.workEnd;
  } else {
    final totalPeriods = totalDays / pd.daysPerPeriod;
    piw = ((totalPeriods.floor() % pd.periodsPerWeek) + pd.periodsPerWeek) % pd.periodsPerWeek;
    isWorkPeriod = piw < pd.workPeriodsPerWeek;
    isWorkHour = isWorkPeriod && localHour >= pd.workStart && localHour < pd.workEnd;
  }

  final yearLenDays = pd.siderealYrMs / solarDay;
  final yearNumber = (totalDays / yearLenDays).floor();
  final dayInYear = (totalDays - yearNumber * yearLenDays).floor();

  int? solInYear;
  int? solsPerYear;
  if (effective == Planet.mars) {
    solInYear = dayInYear;
    solsPerYear = (pd.siderealYrMs / solarDay + 0.5).floor();
  }

  final hStr = h.toString().padLeft(2, '0');
  final mStr = m.toString().padLeft(2, '0');
  final sStr = s.toString().padLeft(2, '0');

  return PlanetTime(
    hour: h,
    minute: m,
    second: s,
    localHour: localHour,
    dayFraction: dayFrac,
    dayNumber: dayNumber,
    dayInYear: dayInYear,
    yearNumber: yearNumber,
    periodInWeek: piw,
    isWorkPeriod: isWorkPeriod,
    isWorkHour: isWorkHour,
    timeStr: '$hStr:$mStr',
    timeStrFull: '$hStr:$mStr:$sStr',
    solInYear: solInYear,
    solsPerYear: solsPerYear,
  );
}

// ── Mars Coordinated Time ─────────────────────────────────────────────────────

/// Get Mars Coordinated Time (MTC) — the Martian equivalent of UTC.
MtcResult getMtc(int utcMs) {
  final totalSols = (utcMs - marsEpochMs) / marsSolMs.toDouble();
  final sol = totalSols.floor();
  final frac = totalSols - sol;

  final h = (frac * 24).floor();
  final m = ((frac * 24 - h) * 60).floor();
  final s = (((frac * 24 - h) * 60 - m) * 60).floor();

  final hStr = h.toString().padLeft(2, '0');
  final mStr = m.toString().padLeft(2, '0');

  return MtcResult(
    sol: sol,
    hour: h,
    minute: m,
    second: s,
    mtcStr: '$hStr:$mStr',
  );
}

File: swift/planet-time/Sources/InterplanetTime/TimeCalc.swift
// TimeCalc.swift — Planet time calculations ported from planet-time.js.

import Foundation

public extension InterplanetTime {

    struct PlanetTime {
        public let hour, minute, second: Int
        public let localHour, dayFraction: Double
        public let dayNumber, dayInYear, yearNumber: Int64
        public let periodInWeek: Int
        public let isWorkPeriod, isWorkHour: Bool
        public let timeStr, timeStrFull: String  // "HH:MM", "HH:MM:SS"
        public let solInYear, solsPerYear: Int64? // Mars only
        public let zoneId: String?               // e.g. "AMT+4"; nil for Earth
    }

    struct MTC {
        public let sol: Int64
        public let hour, minute, second: Int
        public let mtcStr: String // "HH:MM"
    }

    private static let zonePrefix: [String: String] = [
        "mars":    "AMT",
        "moon":    "LMT",
        "mercury": "MMT",
        "venus":   "VMT",
        "jupiter": "JMT",
        "saturn":  "SMT",
        "uranus":  "UMT",
        "neptune": "NMT",
    ]

    static func getPlanetTime(_ planet: String, _ utcMs: Int64, tzOffsetH: Double = 0) -> PlanetTime {
        let effective = planet == "moon" ? "earth" : planet
        let pd = planetData[effective]!
        let solarDay = Double(pd.solarDayMs)

        let elapsedMs  = Double(utcMs - pd.epochMs) + tzOffsetH / 24.0 * solarDay
        let totalDays  = elapsedMs / solarDay
        let dayNumber  = Int64(floor(totalDays))
        let dayFrac    = totalDays - Double(dayNumber)
        let localHour  = dayFrac * 24.0
        let hour       = Int(localHour)
        let minF       = (localHour - Double(hour)) * 60.0
        let minute     = Int(minF)
        let second     = Int((minF - Double(minute)) * 60.0)

        // Work period (positive modulo)
        let piw: Int
        let isWorkPeriod: Bool
        let isWorkHour: Bool
        if pd.earthClockSched {
            // Mercury/Venus: solar day >> circadian rhythm; use UTC Earth-clock scheduling
            // UTC day-of-week: ((floor(utcMs / 86400000) % 7 + 10) % 7 → Mon=0..Sun=6
            // (+7 before +3 ensures positive result for pre-1970 timestamps)
            let utcDay = Int64(floor(Double(utcMs) / Double(86400_000)))
            piw = Int(((utcDay % 7) + 10) % 7)
            isWorkPeriod = piw < pd.workPeriodsPerWeek
            // UTC hour within the day — positive modulo handles pre-1970 timestamps
            let msInDay = ((utcMs % 86400_000) + 86400_000) % 86400_000
            let utcHour = Double(msInDay) / 3_600_000.0
            isWorkHour = isWorkPeriod && utcHour >= Double(pd.workStart) && utcHour < Double(pd.workEnd)
        } else {
            let totalPeriods = totalDays / pd.daysPerPeriod
            piw = ((Int(floor(totalPeriods)) % pd.periodsPerWeek) + pd.periodsPerWeek) % pd.periodsPerWeek
            isWorkPeriod = piw < pd.workPeriodsPerWeek
            isWorkHour = isWorkPeriod && localHour >= Double(pd.workStart) && localHour < Double(pd.workEnd)
        }

        let yearLenDays = Double(pd.siderealYrMs) / solarDay
        let yearNumber  = Int64(floor(totalDays / yearLenDays))
        let dayInYear   = Int64(floor(totalDays - Double(yearNumber) * yearLenDays))

        let solInYear:   Int64? = effective == "mars" ? dayInYear : nil
        let solsPerYear: Int64? = effective == "mars" ? Int64((Double(pd.siderealYrMs) / solarDay).rounded()) : nil

        var zoneId: String? = nil
        if planet != "earth", let prefix = zonePrefix[planet] {
            let off = Int(tzOffsetH)
            let sign = off < 0 ? "-" : "+"
            zoneId = "\(prefix)\(sign)\(abs(off))"
        }

        let h2 = String(format: "%02d", hour)
        let m2 = String(format: "%02d", minute)
        let s2 = String(format: "%02d", second)

        return PlanetTime(
            hour: hour, minute: minute, second: second,
            localHour: localHour, dayFraction: dayFrac,
            dayNumber: dayNumber, dayInYear: dayInYear, yearNumber: yearNumber,
            periodInWeek: piw, isWorkPeriod: isWorkPeriod, isWorkHour: isWorkHour,
            timeStr: "\(h2):\(m2)", timeStrFull: "\(h2):\(m2):\(s2)",
            solInYear: solInYear, solsPerYear: solsPerYear,
            zoneId: zoneId
        )
    }

    static func getMTC(_ utcMs: Int64) -> MTC {
        let ms     = Double(utcMs - marsEpochMs)
        let sol    = Int64(floor(ms / Double(marsSolMs)))
        var fracMs = ms.truncatingRemainder(dividingBy: Double(marsSolMs))
        if fracMs < 0 { fracMs += Double(marsSolMs) }
        let totalSec = fracMs / 1000.0
        let hour   = Int(totalSec / 3600.0)
        let minute = Int(totalSec.truncatingRemainder(dividingBy: 3600.0) / 60.0)
        let second = Int(totalSec.truncatingRemainder(dividingBy: 60.0))
        return MTC(sol: sol, hour: hour, minute: minute, second: second,
                   mtcStr: String(format: "%02d:%02d", hour, minute))
    }
}
